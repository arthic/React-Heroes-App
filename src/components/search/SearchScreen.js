import React, { useMemo } from 'react'
import queryString from 'query-string'

import { HeroCard } from '../heroes/HeroCard'
import { useForm } from '../../hooks/useForm'
import { useLocation } from 'react-router-dom'
import { getHeroByName } from '../../selectors/getHeroByName'

export const SearchScreen = ({history}) => {

	// hook de React-router para menejar props del url
	const location = useLocation()

	// npm install query-string
	const {q = ''} = queryString.parse(location.search)
	// customHook
	const [formValues, handleInputChange] = useForm({
		// manejamos el input con el nombre asigando
		// Se mantiene el texto del input con la url ya parseada
		searchText: q
	})

	// Desestructuramos el searchText para menejarlo en otros lados
	const {searchText} = formValues

	/*Hacer que se dispare la busqueda solo cuando el query
	cambie, osea cuande se de enter*/
	const heroesFiltered = useMemo(() =>
		getHeroByName(q),
		[q]
	)
	// const heroesFiltered = getHeroByName(searchText)

	// Manejo del envio del formulario
	const handleSearch = (e) => {
		e.preventDefault()
		// console.log(searchText);
		// Mandamos la busqueda al query param
		history.push(`?q=${searchText}`)
	}

	return (
		<div>
			<h1 className="publisher">BÃºsqueda</h1>
			<hr/>

			<div className="row animate__animated animate__fadeIn">
				<div className="col-5">
					<hr/>

					<form onSubmit={handleSearch}>
						<input
							type="text"
							placeholder="Busca un Heroe"
							className="form-control"
							name="searchText"
							value={searchText}
							onChange={handleInputChange}
							autoComplete="off"
						/>
						<button
							type="submit"
							className="btn m-1 btn-block btn-outline-primary"
						>	Buscar
						</button>
					</form>
				</div>

				<div className="col-7">
					<hr/>
					{	(q === '')
						&&
						<div className="spinner-border" role="status">
							<span className="sr-only">Loading...</span>
						</div>
					}
					{	(q !== '' && heroesFiltered.length === 0)
						&&
						<div className="alert alert-danger">
							"{q}" No es nombre de Heroe valido :(
						</div>
					}
					{
						heroesFiltered.map(hero => (
							<HeroCard
								key={hero.id}
								{...hero}
							/>
						))
					}
				</div>

			</div>
		</div>
	)
}
