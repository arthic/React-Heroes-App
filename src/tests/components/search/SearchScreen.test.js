import '@testing-library/jest-dom'
import React from 'react';
import {mount} from 'enzyme';
import { SearchScreen } from '../../../components/search/SearchScreen';
import { MemoryRouter, Route } from 'react-router-dom';

describe('Testing en <SearchScreen />', () => {
	test('Mostrarse correctamente con valores por defecto', () => {
		const wrapper = mount(
			<MemoryRouter initialEntries={['/search']}>
				<Route path="/search" component={SearchScreen} />
			</MemoryRouter>
		)
		expect(wrapper).toMatchSnapshot()
		expect(wrapper.find('.spinner-border').exists()).toBe(true)
	})
	test('Debe mostrar a Batman y el input con el valor del queryString', () => {
		const wrapper = mount(
			<MemoryRouter initialEntries={['/search?q=batman']}>
				<Route path="/search" component={SearchScreen} />
			</MemoryRouter>
		)
		expect(wrapper.find('input').prop('value')).toBe('batman')
		expect(wrapper).toMatchSnapshot()
	})
	test('Debe mostrar un error si no se encuentra el Hero', () => {
		const wrapper = mount(
			<MemoryRouter initialEntries={['/search?q=batman123']}>
				<Route path="/search" component={SearchScreen} />
			</MemoryRouter>
		)
		expect(wrapper.find('.alert-danger').text().trim()).toBe('"batman123" No es nombre de Heroe valido :(')
		expect(wrapper).toMatchSnapshot()
	})
	test('Debe de llamar el push del history', () => {
		const history = {
			push: jest.fn()
		}
		const wrapper = mount(
			<MemoryRouter initialEntries={['/search?q=batman123']}>
				<Route
					path="/search"
					component={() => <SearchScreen history={history}/>}
				/>
			</MemoryRouter>
		)
		wrapper.find('input').simulate('change', {
			target: {
				name: 'searchText',
				value: 'batman'
			}
		})
		wrapper.find('form').prop('onSubmit')({
			preventDefault(){}
		})
		expect(history.push).toHaveBeenLastCalledWith('?q=batman')
	});
});
