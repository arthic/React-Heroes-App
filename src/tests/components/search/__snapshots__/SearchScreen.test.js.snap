// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing en <SearchScreen /> Debe mostrar a Batman y el input con el valor del queryString 1`] = `
<div>
  <h1
    className="publisher"
  >
    Búsqueda
  </h1>
  <hr />
  <div
    className="row animate__animated animate__fadeIn"
  >
    <div
      className="col-5"
    >
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Busca un Heroe"
          type="text"
          value="batman"
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
           Buscar
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <hr />
      <div
        className="card ms-3 animate__animated animate__fadeIn"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Batman"
              className="card-img"
              src="./assets/heroes/dc-batman.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Detective Comics #27
                </small>
              </p>
              <a
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                Más...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Testing en <SearchScreen /> Debe mostrar un error si no se encuentra el Hero 1`] = `
<div>
  <h1
    className="publisher"
  >
    Búsqueda
  </h1>
  <hr />
  <div
    className="row animate__animated animate__fadeIn"
  >
    <div
      className="col-5"
    >
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Busca un Heroe"
          type="text"
          value="batman123"
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
           Buscar
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <hr />
      <div
        className="alert alert-danger"
      >
        "
        batman123
        " No es nombre de Heroe valido :(
      </div>
    </div>
  </div>
</div>
`;

exports[`Testing en <SearchScreen /> Mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1
    className="publisher"
  >
    Búsqueda
  </h1>
  <hr />
  <div
    className="row animate__animated animate__fadeIn"
  >
    <div
      className="col-5"
    >
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Busca un Heroe"
          type="text"
          value=""
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
           Buscar
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <hr />
      <div
        className="spinner-border"
        role="status"
      >
        <span
          className="sr-only"
        >
          Loading...
        </span>
      </div>
    </div>
  </div>
</div>
`;
